snowflake/src/main/java/org/apache/iceberg/snowflake/SnowflakeIdentifier.java

```diff
@@ -116,16 +119,46 @@ public int hashCode() {
   public String toIdentifierString() {
     switch (type()) {
       case TABLE:
-        return String.format("%s.%s.%s", databaseName, schemaName, tableName);
+        return String.format(
+            "%s.%s.%s",
+            toIdentifierString(databaseName),
+            toIdentifierString(schemaName),
+            toIdentifierString(tableName));
       case SCHEMA:
-        return String.format("%s.%s", databaseName, schemaName);
+        return String.format(
+            "%s.%s", toIdentifierString(databaseName), toIdentifierString(schemaName));
       case DATABASE:
-        return databaseName;
+        return toIdentifierString(databaseName);
       default:
         return "";
     }
   }
 
+  /**
+   * Returns the supplied identifier String as a String suitable for use in a Snowflake IDENTIFIER
+   * param.
+   */
+  @VisibleForTesting
+  static String toIdentifierString(String identifier) {
+    // Identifier with special characters are case-sensitive while regular identifiers (no special
+    // characters) are
+    // case-insensitive. More details could be found here.
+    // https://docs.snowflake.com/en/sql-reference/identifiers-syntax.html
+    //
+    Pattern noSpecialChars = Pattern.compile("[^a-z0-9]", Pattern.CASE_INSENSITIVE);
```

Comment:
As discussed, this regex isn't quite right, but also we should check whether our purpose is really to escape special characters directly, or to allow the user-input to plumb though additional quotes themselves.

For example, create database "mIxEd" doesn't have "special" characters and yet the only way for the user to plumb that all the way through is if "escaped" quotes make it all the way through. The spark-side syntax might have to have a second layer like:

    spark.sql("show namespaces in "mIxEd"");

And we'd have to replace \" with \"\", but we don't have to additionally wrap the whole thing in our own double-quotes.

Once we have a the full set of test cases thought out, I wonder if something as simple as replacing all of this with:

     return identifier.replaceAll("\"", "\"\"");

achieves what we want.

Reply:
Removed all the special character handling code from the catalog. Will rely completely on the user input to interpret if the identifier is quoted or unquoted. I have added examples in the PR description about how different cases would be handled.